### ISSUE TITLE
new requirement: support for filter one function itself instead of whole subtree.

### ISSUE THREAD

**Issue by simpzan** [namhyung/uftrace#1169]
I am tracing wayfire, a window compositor for linux. the tracing log is very huge for just 20 second. and a large part of the log is from std template. I tried to filter out all std function with `-F std::.`, it turns out most user functions are filtered out as well. because some user function is called in the subtree of a std function like cpp lambda. 
I just want to filter out all std function and keep all user functions, how to achieve this? one feature to achieve this is using filter based on one function instead of a whole subtree, but the feature seems not available in uftrace now. Any advice? thanks!

**Comment by namhyung** [comment]
Hi, thanks for using uftrace.

The `-F std::.` will show std functions (and their subtrees) only, what about `-N std::.` instead?

**Comment by simpzan** [comment]
Oh, sorry, I made a mistake. I do use -N std::. and it filtered out all std
functions, meanwhile many user functions in the sub tree of std functions
are also filtered out. For example, user functions called from cpp
 lambda.  That’s not what I want. I want to hide all std functions and keep
all user functions.

**Comment by namhyung** [comment]
OK, we don't have such filter just for single function.

Instead, you can try depth trigger like `-T std::.@depth=0`, it'll hide all functions started with std namesapce.  But you need to restore depth for user functions manually.  If your program starts with a common prefix or contains some unique name, you can use `-T prefix@depth=100` for example.

**Comment by namhyung** [comment]
So I've played with it. Please test [check/hide-filter](https://github.com/namhyung/uftrace/compare/check/hide-filter).  `--hide` option was added for testing.

Normal run:
```
$ uftrace a.out | head
# DURATION     TID     FUNCTION
            [ 86842] | main() {
            [ 86842] |   Map::Map() {
            [ 86842] |     std::unordered_map::unordered_map() {
            [ 86842] |       std::_Hashtable::_Hashtable() {
            [ 86842] |         std::__detail::_Hashtable_alloc::_Hashtable_alloc() {
            [ 86842] |           std::__detail::_Hashtable_ebo_helper::_Hashtable_ebo_helper() {
            [ 86842] |             std::allocator::allocator() {
   0.230 us [ 86842] |               __gnu_cxx::new_allocator::new_allocator();
   2.549 us [ 86842] |             } /* std::allocator::allocator */
```

To hide std library functions:
```
$ uftrace --hide ^std a.out | head
# DURATION     TID     FUNCTION
            [ 86895] | main() {
            [ 86895] |   Map::Map() {
   0.105 us [ 86895] |     __gnu_cxx::new_allocator::new_allocator();
   4.568 us [ 86895] |   } /* Map::Map */
            [ 86895] |   Map::AddTask() {
   1.930 us [ 86895] |     malloc();
   1.008 us [ 86895] |     strdup();
            [ 86895] |     __gnu_cxx::new_allocator::allocate() {
   0.062 us [ 86895] |       __gnu_cxx::new_allocator::max_size();
```

**Comment by honggyukim** [comment]
I've been thinking about this feature so created issues before at #898. If `-U` option works for `-pg` compiled binaries, then it will work exactly same as `--hide` filter above.

#1149 can be supported for analysis commands and it will provide consistent filter interface along with #898.

@simpzan Thanks for bringing attention to this feature.

**Comment by simpzan** [comment]
@namhyung @honggyukim Thank you for quick response. It's amazing. 

@namhyung I tested the new branch. `live` is working nicely, while `replay` and `dump --chrome` is not. 

> So I've played with it. Please test [check/hide-filter](https://github.com/namhyung/uftrace/compare/check/hide-filter). `--hide` option was added for testing.
> 
> Normal run:
> 
> ```
> $ uftrace a.out | head
> # DURATION     TID     FUNCTION
>             [ 86842] | main() {
>             [ 86842] |   Map::Map() {
>             [ 86842] |     std::unordered_map::unordered_map() {
>             [ 86842] |       std::_Hashtable::_Hashtable() {
>             [ 86842] |         std::__detail::_Hashtable_alloc::_Hashtable_alloc() {
>             [ 86842] |           std::__detail::_Hashtable_ebo_helper::_Hashtable_ebo_helper() {
>             [ 86842] |             std::allocator::allocator() {
>    0.230 us [ 86842] |               __gnu_cxx::new_allocator::new_allocator();
>    2.549 us [ 86842] |             } /* std::allocator::allocator */
> ```
> 
> To hide std library functions:
> 
> ```
> $ uftrace --hide ^std a.out | head
> # DURATION     TID     FUNCTION
>             [ 86895] | main() {
>             [ 86895] |   Map::Map() {
>    0.105 us [ 86895] |     __gnu_cxx::new_allocator::new_allocator();
>    4.568 us [ 86895] |   } /* Map::Map */
>             [ 86895] |   Map::AddTask() {
>    1.930 us [ 86895] |     malloc();
>    1.008 us [ 86895] |     strdup();
>             [ 86895] |     __gnu_cxx::new_allocator::allocate() {
>    0.062 us [ 86895] |       __gnu_cxx::new_allocator::max_size();
> ```

**Comment by namhyung** [comment]
@honggyukim yes, I think `-P` and `-U` would have a different semantics and will be used for dynamic tracing.  While `--hide` is worked as a filter like `-N`.

@simpzan I only implemented the record part.  After you record it with `--hide`, you can use it as usual (or even apply other filters like `-F` or `-N`).

**Comment by simpzan** [comment]
@namhyung Oh, I see. I tested it, and it works. Thanks! 

In current implementation, I have to record twice to get 2 versions of traces, one with the std functions and one without the std functions. Is it possible to record just once and use `replay` and `dump` to generate 2 versions of traces? I think it will be much easier to use this way. Thanks.

**Comment by honggyukim** [comment]
@simpzan uftrace currently doesn't support to regenerate a new trace based on the existing record. I recommend you to record once then replay with filters that you prefer.

If you feel it's inconvenient to give filters for every replay execution, there might be another way to do this with `uftrace.data/default.opts`. If you add options to this file for replay or dump, then the options will be applied by default for the data record. So you can just copy the same record data, but add different filters in `default.opts` to each data record.

**Comment by simpzan** [comment]
@honggyukim thanks for your response. I think there's a misunderstanding here. Actually I do want to record once and replay with different filters later. but the current implementation of `--hide` don't seems to support that. if I record without `--hide` filters, then replay with `--hide` filters, the filters won't work.

currently the only way to use `--hide` filters is to use it with `record` subcommand. so in order to get 2 trace logs, I have to record twice. but sometimes it's not easy to record twice with same program behaviors due to some random variables in the program.

**Comment by honggyukim** [comment]
@simpzan Please test [check/hide-filter-v2](https://github.com/namhyung/uftrace/compare/check/hide-filter-v2).  `--hide` option is now working for both record and replay.

For example, `t-abc` program can be compiled and recorded as follows:
```
$ gcc -pg -o t-abc tests/s-abc.c
$ uftrace record --hide b t-abc
```
It doesn't show `b()` function as expected in replay output.
```
$ uftrace replay
# DURATION     TID     FUNCTION
  31.300 us [ 12131] | __monstartup();
   2.200 us [ 12131] | __cxa_atexit();
            [ 12131] | main() {
            [ 12131] |   a() {
            [ 12131] |     c() {
   2.700 us [ 12131] |       getpid();
   5.800 us [ 12131] |     } /* c */
  12.000 us [ 12131] |   } /* a */
  15.100 us [ 12131] | } /* main */
```
The `--hide` option can also be used in replay time to hide `c()` function for the same record.
```
$ uftrace replay --hide c
# DURATION     TID     FUNCTION
  31.300 us [ 12131] | __monstartup();
   2.200 us [ 12131] | __cxa_atexit();
            [ 12131] | main() {
            [ 12131] |   a() {
   2.700 us [ 12131] |     getpid();
  12.000 us [ 12131] |   } /* a */
  15.100 us [ 12131] | } /* main */
```
The function `c()` can be hidden with a trigger option as follows:
```
$ uftrace replay -T c@hide
# DURATION     TID     FUNCTION
  31.300 us [ 12131] | __monstartup();
   2.200 us [ 12131] | __cxa_atexit();
            [ 12131] | main() {
            [ 12131] |   a() {
   2.700 us [ 12131] |     getpid();
  12.000 us [ 12131] |   } /* a */
  15.100 us [ 12131] | } /* main */
```

**Comment by honggyukim** [comment]
> @honggyukim yes, I think -P and -U would have a different semantics and will be used for dynamic tracing. While --hide is worked as a filter like -N.

@namhyung I still think that `-U` can be used for this purpose to provide a consistent filter interface for both record and replay time. The `-U` option can unpatch if the target function is dynamically patched, but otherwise it can be used same as `--hide` filter.

Similarly, `-P` option can also be used as a filter without showing subtree for normal functions as well.

**Comment by simpzan** [comment]
@honggyukim I just tested the new patch. everything is working as expected. Thank you! 

I have to say that uftrace is really a wonderful tool to help understand large unfamiliar codebase. Thank you for creating/maintaining this amazing tool. :)  @namhyung @honggyukim 

> @simpzan Please test [check/hide-filter-v2](https://github.com/namhyung/uftrace/compare/check/hide-filter-v2). `--hide` option is now working for both record and replay.
> 
> For example, `t-abc` program can be compiled and recorded as follows:
> 
> ```
> $ gcc -pg -o t-abc tests/s-abc.c
> $ uftrace record --hide b t-abc
> ```
> 
> It doesn't show `b()` function as expected in replay output.
> 
> ```
> $ uftrace replay
> # DURATION     TID     FUNCTION
>   31.300 us [ 12131] | __monstartup();
>    2.200 us [ 12131] | __cxa_atexit();
>             [ 12131] | main() {
>             [ 12131] |   a() {
>             [ 12131] |     c() {
>    2.700 us [ 12131] |       getpid();
>    5.800 us [ 12131] |     } /* c */
>   12.000 us [ 12131] |   } /* a */
>   15.100 us [ 12131] | } /* main */
> ```
> 
> The `--hide` option can also be used in replay time to hide `c()` function for the same record.
> 
> ```
> $ uftrace replay --hide c
> # DURATION     TID     FUNCTION
>   31.300 us [ 12131] | __monstartup();
>    2.200 us [ 12131] | __cxa_atexit();
>             [ 12131] | main() {
>             [ 12131] |   a() {
>    2.700 us [ 12131] |     getpid();
>   12.000 us [ 12131] |   } /* a */
>   15.100 us [ 12131] | } /* main */
> ```
> 
> The function `c()` can also be hidden with a trigger option as follows:
> 
> ```
> $ uftrace replay -T c@hide
> # DURATION     TID     FUNCTION
>   31.300 us [ 12131] | __monstartup();
>    2.200 us [ 12131] | __cxa_atexit();
>             [ 12131] | main() {
>             [ 12131] |   a() {
>    2.700 us [ 12131] |     getpid();
>   12.000 us [ 12131] |   } /* a */
>   15.100 us [ 12131] | } /* main */
> ```

**Comment by honggyukim** [comment]
@simpzan Thanks very much for the compliment. I hope it will be helpful to you anytime you need. 

Your feedback was helpful for us to make uftrace better. It'd be great if you could give us more feedback while using uftrace. Thanks a lot.

**Comment by namhyung** [comment]
@honggyukim OK, let's just use `-P` and `-U` instead of adding a new `--hide` option.

**Comment by honggyukim** [comment]
I've uploaded a PR for this that reuses -U option to replace --hide option.

**Comment by honggyukim** [comment]
@simpzan Thanks for your request. It's quite late but the `--hide` option is merged from #1217.
